src/uv_append.c

```diff
@@ -952,13 +950,22 @@ static void uvBarrierClose(struct uv *uv)
         segment->barrier = NULL;
     }
 
-    /* There might still still be a current barrier set on uv->barrier, meaning
+    /* There might still be a current barrier set on uv->barrier, meaning
      * that the open segment it was associated with has started to be finalized
-     * and is not anymore in the append_segments queue. Let's cancel that
-     * too. */
+     * and is not anymore in the append_segments queue. Let's cancel all
+     * untriggered barrier request callbacks too. */
     if (uv->barrier != NULL) {
         uvBarrierTriggerAll(uv->barrier);
-        uv->barrier = NULL;
+        /* Clear uv->barrier if there's no active work on the thread pool. When
+         * the work on the threadpool finishes, UvUnblock will notice
+         * we're closing and call uvMaybeFireCloseCb. UnUnblock will not try to
```

Comment:
Will add in comment that UvUnblock will free and clear UvBarrier, but I'm not force pushing while there's a review in progress.